prov/rxm/src/rxm_cq.c

```diff
@@ -1382,7 +1382,17 @@ void rxm_ep_do_progress(struct util_ep *util_ep)
 	}
 
 	do {
+		/* Comply with restricted threading levels that MSG provider
+		 * may have been configured with. cmap->acquire would be no-op
+		 * in case rxm data progress is manual */
+		if (rxm_ep->msg_info->domain_attr->threading != FI_THREAD_SAFE)
+			rxm_ep->cmap->acquire(&rxm_ep->cmap->lock);
```

Comment:
It is really not clear why we need to cmap lock here, particularly, since we're only reading the CQ (call to another provider) and not actually doing any processing while holding the lock.  So, not only does this look like the wrong lock, but it also looks unnecessary.

Reply:
The atomics auto-progress thread would read MSG CQ as part of fi_trywait. In case we have configured the MSG provider with a threading level other than FI_THREAD_SAFE we need to serialize access to MSG CQ. We're just re-using cmap->lock to do that. Would defining a new lock for this purpose make things clear?

Reply:
A new lock would help.  Locking the cmap doesn't look right.  Could the ep lock be used instead?  Or maybe you need a per rxm ep msg_resource_lock?

Part of the confusion is that here you're using the cmap lock to protect access to a resource associated with an EP.  But above (see my other comment), you use the EP lock to protect access to cmap data.  Something looks off.

Reply:
Over time, the use case of cmap->lock expanded to protecting MSG provider resources as well. I think there was a change added last year which used cmap->lock to guarantee single threaded acccess of libibverbs resources to improve performance. There's a libibverbs env variable which when set would remove all locks in the library.

ep->lock can't be used because it can be no-op when threading level is not FI_THREAD_SAFE and we still have a progress thread in rxm. We could have a msg_resource_lock.

Reply:
I would go with msg_resource_lock, associated it with the ep, and remove the cmap->lock if that works.

Reply:
I have one concern with defining a new msg_resource_lock. We would need to ensure correct lock ordering in case we have to acquire both cmap->lock and msg_resource_lock. I can see this happening when we close a msg ep that failed a simultaneous connection attempt.

As an alternative, would it help if we rename cmap_lock to cm/progress_thread_lock or something like that?

Reply:
I was suggesting removing the lock from struct rxm_cmap.  Use an EP related lock (new msg_resource_lock) to protect access to it.  There's a 1:1 relationship between an ep and cmap, correct?  Do we technically even need a separate struct for cmap, versus embedding the fields in the EP?  (I'm not suggesting to do this.)